image/refs/interface.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package refs implements generic name-based reference access.
+package refs
+
+import (
+	"github.com/opencontainers/image-spec/specs-go"
+	"golang.org/x/net/context"
+)
+
+// Engine represents a name-based reference storage engine.
+//
+// This interface is for internal use of oci-image-tool for the time
+// being.  It is subject to change.  This notice will be removed when
+// and if the interface becomes stable.
+type Engine interface {
+
+	// Put adds a new reference to the store.  The action is idempotent;
+	// a nil return means "that descriptor is stored at NAME" without
+	// implying "because of your Put()".
+	Put(ctx context.Context, name string, descriptor *specs.Descriptor) (err error)
```

Comment:
On Mon, Aug 29, 2016 at 06:59:21PM -0700, Stephen Day wrote:
> > +	// Put adds a new reference to the store.  The action is idempotent;
> > +	// a nil return means "that descriptor is stored at NAME" without
> > +	// implying "because of your Put()".
> > +	Put(ctx context.Context, name string, descriptor *specs.Descriptor) (err error)
> 
> Don't pass the `Descriptor` type as a pointer type. Treat as a value.

Why?  It seems like we might as well save a copy operation, although
Descriptor isn't very big.  Is there concern about error-checking nil
values?  Or that Put somehow mutates the passed-in descriptor?  Or
some other reason?